File: src/tools/execute.ts
import { terminalManager } from '../terminal-manager.js';
import { commandManager } from '../command-manager.js';
import { logToFile } from '../logging.js';
import { ExecuteCommandArgsSchema, ReadOutputArgsSchema, ForceTerminateArgsSchema, ListSessionsArgsSchema } from './schemas.js';

export async function executeCommand(args: unknown) {
  await logToFile('Processing execute_command request');
  const parsed = ExecuteCommandArgsSchema.safeParse(args);
  if (!parsed.success) {
    throw new Error(`Invalid arguments for execute_command: ${parsed.error}`);
  }

  if (!commandManager.validateCommand(parsed.data.command)) {
    throw new Error(`Command not allowed: ${parsed.data.command}`);
  }

  const result = await terminalManager.executeCommand(
    parsed.data.command,
    parsed.data.timeout_ms
  );
  
  await logToFile(`Executed command: ${parsed.data.command} with PID ${result.pid}`);
  return {
    content: [{
      type: "text",
      text: `Command started with PID ${result.pid}\nInitial output:\n${result.output}${
        result.isBlocked ? '\nCommand is still running. Use read_output to get more output.' : ''
      }`
    }],
  };
}

export async function readOutput(args: unknown) {
  await logToFile('Processing read_output request');
  const parsed = ReadOutputArgsSchema.safeParse(args);
  if (!parsed.success) {
    throw new Error(`Invalid arguments for read_output: ${parsed.error}`);
  }

  const output = terminalManager.getNewOutput(parsed.data.pid);
  return {
    content: [{
      type: "text",
      text: output === null
        ? `No session found for PID ${parsed.data.pid}`
        : output || 'No new output available'
    }],
  };
}

export async function forceTerminate(args: unknown) {
  await logToFile('Processing force_terminate request');
  const parsed = ForceTerminateArgsSchema.safeParse(args);
  if (!parsed.success) {
    throw new Error(`Invalid arguments for force_terminate: ${parsed.error}`);
  }

  const success = terminalManager.forceTerminate(parsed.data.pid);
  return {
    content: [{
      type: "text",
      text: success
        ? `Successfully initiated termination of session ${parsed.data.pid}`
        : `No active session found for PID ${parsed.data.pid}`
    }],
  };
}

export async function listSessions() {
  await logToFile('Processing list_sessions request');
  const sessions = terminalManager.listActiveSessions();
  return {
    content: [{
      type: "text",
      text: sessions.length === 0
        ? 'No active sessions'
        : sessions.map(s =>
            `PID: ${s.pid}, Blocked: ${s.isBlocked}, Runtime: ${Math.round(s.runtime / 1000)}s`
          ).join('\n')
    }],
  };
}

File: py/desktop_commander/tools/execute.py
from typing import Dict, Any, List
from desktop_commander.terminal_manager import terminal_manager
from desktop_commander.command_manager import command_manager
from desktop_commander.tools.schemas import (
    ExecuteCommandArgs,
    ReadOutputArgs,
    ForceTerminateArgs,
    ListSessionsArgs
)
from desktop_commander.types import ToolResponse


async def execute_command(args: ExecuteCommandArgs) -> ToolResponse:
    """Execute a terminal command."""
    # Validate command is allowed
    if not command_manager.validate_command(args.command):
        raise Exception(f"Command not allowed: {args.command}")
    
    # Execute command
    result = await terminal_manager.execute_command(
        args.command,
        args.timeout_ms if args.timeout_ms is not None else None
    )
    
    message = (
        f"Command started with PID {result.pid}\n"
        f"Initial output:\n{result.output}"
    )
    
    if result.is_blocked:
        message += "\nCommand is still running. Use read_output to get more output."
    
    return ToolResponse(
        content=[{"type": "text", "text": message}]
    )


async def read_output(args: ReadOutputArgs) -> ToolResponse:
    """Read new output from a running terminal session."""
    output = terminal_manager.get_new_output(args.pid)
    
    message = (
        "No session found for PID {args.pid}" if output is None
        else output or "No new output available"
    )
    
    return ToolResponse(
        content=[{"type": "text", "text": message}]
    )


async def force_terminate(args: ForceTerminateArgs) -> ToolResponse:
    """Force terminate a running terminal session."""
    success = terminal_manager.force_terminate(args.pid)
    
    message = (
        f"Successfully initiated termination of session {args.pid}" if success
        else f"No active session found for PID {args.pid}"
    )
    
    return ToolResponse(
        content=[{"type": "text", "text": message}]
    )


async def list_sessions() -> ToolResponse:
    """List all active terminal sessions."""
    sessions = terminal_manager.list_active_sessions()
    
    if not sessions:
        message = "No active sessions"
    else:
        message = "\n".join(
            f"PID: {s.pid}, Blocked: {s.is_blocked}, "
            f"Runtime: {round(s.runtime / 1000)}s"
            for s in sessions
        )
    
    return ToolResponse(
        content=[{"type": "text", "text": message}]
    )